src/cli.rs

```diff
@@ -1,3 +1,354 @@
-pub fn run() {
-    println!("Hello, world!");
+#[cfg(test)]
+use mockall::automock;
+
+use crate::adapters::operations::device::DeviceOperations;
+use crate::core::operations::Operations;
+use crate::devices::local_file_storage::{
+    LocalFileStorage, StandardFileSystem, StandardPathProvider,
+};
+
+const HELP: &str = r#"
+HibernaCLI
+Usage:
+    hibernacli [command] [options]
+
+Commands:
+    help              Display this help message
+    --version or -v   Display the version of the application
+    device [opt]      Manage devices
+    config [opt]      Manage configuration
+"#;
+
+const INVALID_COMMAND: &str = "Invalid command, use 'help' to display available commands";
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const DEFAULT_CONFIG: &str = r#"[[devices]]
+    name = "MyPersonalDevice"
+    type = "MockDevice"
+"#;
+
+#[cfg_attr(test, automock)]
+trait UserInterface {
+    fn write(&self, message: &str) -> Result<(), String>;
+    fn read(&self) -> Result<String, String>;
+}
+
+struct Console;
+
+impl UserInterface for Console {
+    fn write(&self, message: &str) -> Result<(), String> {
+        println!("{}", message);
+        Ok(())
+    }
+    fn read(&self) -> Result<String, String> {
+        let mut buffer = String::new();
+        match std::io::stdin().read_line(&mut buffer) {
+            Ok(_) => return Ok(buffer),
+            Err(e) => return Err(e.to_string()),
+        };
+    }
+}
+
+struct CommandRunner<T: UserInterface, U: DeviceOperations> {
+    console: T,
+    operations: U,
+}
+
+impl<T: UserInterface, U: DeviceOperations> CommandRunner<T, U> {
+    fn run(&mut self, args: Vec<String>) {
+        if args.len() < 2 {
+            self.display_invalid_command();
+            return;
+        }
+
+        match args[1].as_str() {
+            "help" => self.display_help(),
+            "--version" | "-v" => self.display_version(),
+            "device" => self.run_device_command(args),
+            _ => {
+                self.display_invalid_command();
+            }
+        }
+    }
+
+    fn display_message(&mut self, message: &String) {
+        let _ = self.console.write(message);
+    }
+
+    fn read_string(&mut self) -> Result<String, String> {
+        self.console.read()
+    }
+
+    fn read_number(&mut self) -> Result<i32, String> {
+        let input = self.console.read()?;
+        match input.trim().parse::<i32>() {
+            Ok(number) => Ok(number),
+            Err(_) => Err("Invalid number".to_string()),
+        }
+    }
+
+    fn display_help(&mut self) {
+        let _ = self.display_message(&HELP.to_string());
```

Comment:
Therefore here, you don't need a mutable reference, and you can pass directly `display_message(HELP)`